cpp/src/io/json/column_tree_construction.cu

```diff
@@ -0,0 +1,549 @@
+  // column:
+  //    reduce_by_key {col_id}, {node_categories} - custom opp (*+v=*, v+v=v, *+#=E)
+  rmm::device_uvector<row_offset_t> max_row_offsets(num_columns, stream);
+  rmm::device_uvector<NodeT> column_categories(num_columns, stream);
+  auto ordered_row_offsets =
+    thrust::make_permutation_iterator(row_offsets.begin(), level_ordered_node_ids.begin());
+  auto ordered_node_categories =
+    thrust::make_permutation_iterator(tree.node_categories.begin(), level_ordered_node_ids.begin());
+  thrust::reduce_by_key(
+    rmm::exec_policy_nosync(stream),
+    level_ordered_col_ids.begin(),
+    level_ordered_col_ids.end(),
+    thrust::make_zip_iterator(ordered_row_offsets, ordered_node_categories),
+    thrust::make_discard_iterator(),
+    thrust::make_zip_iterator(max_row_offsets.begin(), column_categories.begin()),
+    thrust::equal_to<NodeIndexT>(),
+    [] __device__(auto a, auto b) {
+      auto row_offset_a = thrust::get<0>(a);
+      auto row_offset_b = thrust::get<0>(b);
+      auto type_a       = thrust::get<1>(a);
+      auto type_b       = thrust::get<1>(b);
+
+      NodeT ctg;
+      auto is_a_leaf = (type_a == NC_VAL || type_a == NC_STR);
+      auto is_b_leaf = (type_b == NC_VAL || type_b == NC_STR);
+      // (v+v=v, *+*=*,  *+v=*, *+#=E, NESTED+VAL=NESTED)
+      // *+*=*, v+v=v
+      if (type_a == type_b) {
+        ctg = type_a;
+      } else if (is_a_leaf) {
+        // *+v=*, N+V=N
+        // STRUCT/LIST + STR/VAL = STRUCT/LIST, STR/VAL + FN = ERR, STR/VAL + STR = STR
+        ctg = (type_b == NC_FN ? NC_ERR : (is_b_leaf ? NC_STR : type_b));
+      } else if (is_b_leaf) {
+        ctg = (type_a == NC_FN ? NC_ERR : (is_a_leaf ? NC_STR : type_a));
+      } else
+        ctg = NC_ERR;
+
+      thrust::maximum<row_offset_t> row_offset_op;
+      return thrust::make_pair(row_offset_op(row_offset_a, row_offset_b), ctg);
+    });
+
+  // 4. construct parent_col_ids using permutation iterator
+  rmm::device_uvector<NodeIndexT> parent_col_ids(num_columns, stream);
+  thrust::transform_output_iterator parent_col_ids_it(
+    parent_col_ids.begin(), parent_nodeids_to_colids{col_ids, rev_mapped_col_ids});
+  thrust::copy_n(rmm::exec_policy_nosync(stream),
+                 thrust::make_permutation_iterator(tree.parent_node_ids.begin(),
+                                                   level_ordered_unique_node_ids.begin()),
+                 num_columns,
+                 parent_col_ids_it);
+
+  /*
+    5. CSR construction:
+      a. Sort column levels and get their ordering
+      b. For each column node coln iterated according to sorted_column_levels; do
+          i. Find nodes that have coln as the parent node -> set adj_coln
+          ii. row idx[coln] = size of adj_coln + 1
+          iii. col idx[coln] = adj_coln U {parent_col_id[coln]}
+  */
+
+  rmm::device_uvector<NodeIndexT> rowidx(num_columns + 1, stream);
+  thrust::fill(rmm::exec_policy_nosync(stream), rowidx.begin(), rowidx.end(), 0);
```

Comment:
you can use `make_zeroed_device_uvector_async` here

Reply:
Replaced!